File: packages/kotlin/src/main/kotlin/com/wave/unlock/UnlockState.kt
package com.wave.unlock

/** States an unlock streams through. Labels match contract/conformance/state-sequences.json. */
sealed class UnlockState(val label: String) {
    object Idle : UnlockState("idle")
    object Scanning : UnlockState("scanning")
    data class ReaderFound(val rssi: Int) : UnlockState("readerFound")
    data class TooFar(val rssi: Int) : UnlockState("tooFar")
    object Writing : UnlockState("writing")
    object AwaitingConfirmation : UnlockState("awaitingConfirmation")
    data class Granted(val reason: String?) : UnlockState("granted")
    data class Denied(val reason: String) : UnlockState("denied")
    object TimedOut : UnlockState("timedOut")
    data class Failed(val error: WaveError) : UnlockState("failed")

    val isTerminal: Boolean
        get() = this is Granted || this is Denied || this is TimedOut || this is Failed
}

enum class WaveError { BLUETOOTH_OFF, PERMISSION_DENIED, WRITE_FAILED, NETWORK, AUTH }

File: packages/swift/Sources/WaveUnlock/UnlockState.swift
import Foundation

/// The states an unlock attempt streams through. Names match the shared conformance
/// vectors (contract/conformance/state-sequences.json).
public enum UnlockState: Equatable, Sendable {
    case idle
    case scanning
    case readerFound(rssi: Int)
    case tooFar(rssi: Int)
    case writing
    case awaitingConfirmation
    case granted(reason: String?)
    case denied(reason: String)
    case timedOut
    case failed(WaveError)

    /// The stable string label used by the conformance vectors.
    public var label: String {
        switch self {
        case .idle: return "idle"
        case .scanning: return "scanning"
        case .readerFound: return "readerFound"
        case .tooFar: return "tooFar"
        case .writing: return "writing"
        case .awaitingConfirmation: return "awaitingConfirmation"
        case .granted: return "granted"
        case .denied: return "denied"
        case .timedOut: return "timedOut"
        case .failed: return "failed"
        }
    }

    public var isTerminal: Bool {
        switch self {
        case .granted, .denied, .timedOut, .failed: return true
        default: return false
        }
    }
}

public enum WaveError: Error, Equatable, Sendable {
    case bluetoothOff
    case permissionDenied
    case bluetoothDisabledByConfig
    case writeFailed(String)
    case network(String)
    case auth(String)
}
